### PR TITLE
Experimental Next.js 12 Output File Tracing

### PR DESCRIPTION
**Pull request by iiroj**
This PR implements the Next.js 12 [Output File Tracing](https://nextjs.org/docs/advanced-features/output-file-tracing), meaning that `@sls-next/serverless-component` will no longer set any `target` while building Next.js, and will instead use the generated `*.nft.json` files to copy required files into the Lambda@Edge deployment.

To use the new feature, one should:

- update to Next.js 12
- make sure there is no `target` in the Next.js configuration (it's now deprecated)
- set the `build.outputFileTracing: true` in the `serverless.yml` inputs for `@sls-next/serverless-component`

### CODE REVIEW

**Comment by iiroj** [comment]
Ping @dphang I replaced https://github.com/serverless-nextjs/serverless-next.js/pull/2124 with this.

**Comment by jonahallibone** [comment]
Is this a no-go? I'd really love to use the new output format

**Comment by iiroj** [comment]
@jonahallibone I haven't had time to write really exhaustive tests... In the meantime, it seems Next.js is already working on a way to do the output themselves: https://nextjs.org/docs/advanced-features/output-file-tracing#automatically-copying-traced-files-experimental

File: jest.config.js
Comment (iiroj) [inline jest.config.js:25]: These were included in the coverage. Shouldn't be too important since they're not available in CI, but they do mess up local coverage.

File: packages/libs/core/src/build/builder.ts
Comment (iiroj) [inline packages/libs/core/src/build/builder.ts:65]: Without the `target` option, it's implied to emit into `.next/server/`.

File: packages/libs/lambda-at-edge/src/lib/copyOutputFileTraces.ts
Comment (iiroj) [inline packages/libs/lambda-at-edge/src/lib/copyOutputFileTraces.ts:20]: This file has some generic default stuff in it, and should always be included.

File: packages/libs/lambda-at-edge/src/lib/copyOutputFileTraces.ts
Comment (iiroj) [inline packages/libs/lambda-at-edge/src/lib/copyOutputFileTraces.ts:25]: Every built "page" has its own nft file.

File: packages/libs/lambda-at-edge/src/lib/copyOutputFileTraces.ts
Comment (iiroj) [inline packages/libs/lambda-at-edge/src/lib/copyOutputFileTraces.ts:47]: After every file from the nft json is added to the `traces` set, we turn it back into an array. This way there are no duplicates!

File: packages/libs/lambda-at-edge/src/lib/copyOutputFileTraces.ts
Comment (iiroj) [inline packages/libs/lambda-at-edge/src/lib/copyOutputFileTraces.ts:48]: There seem to be a lot of `package.json` files, I don't think we need those.

File: packages/libs/lambda-at-edge/src/lib/copyOutputFileTraces.ts
Comment (iiroj) [inline packages/libs/lambda-at-edge/src/lib/copyOutputFileTraces.ts:63]: This is roughly the same logic as in the previous implementation, but I was thinking it would make more sense to generically check that only target paths inside the output directory are included.

Comment (iiroj) [reply packages/libs/lambda-at-edge/src/lib/copyOutputFileTraces.ts:63]: Returning an empty array from `Array.prototype.flatMap` is the same as filtering it out.

File: packages/libs/lambda-at-edge/src/lib/copyRequiredServerFiles.ts
Comment (iiroj) [inline packages/libs/lambda-at-edge/src/lib/copyRequiredServerFiles.ts:14]: Not really sure what the point of this file is, and why it's not an `.nft.json`, but seems to have required content anyway.

**Comment by Purii** [comment]
Awesome! We absolutely need this feature to include localization files.
Is there anything I can support with or is this PR only waiting for approval + merge by @dphang ?

**Comment by iiroj** [comment]
@Purii I do want to write more tests because otherwise the codecoverage check doesn't pass, but it does seem to work at least for my small site!

File: packages/serverless-components/nextjs-component/src/component.ts
Comment (dphang) [inline packages/serverless-components/nextjs-component/src/component.ts:224]: probably can cover it in custom inputs test? or if it's not too convenient, we could modify the existing e2e tests to cover this

**Comment by dphang** [comment]
thanks! I think it looks fine for the most part, codecov drop is just 0.08%, I gave a minor comment, I can probably add a new app to test this later so we get coverage there

**Review by dphang — APPROVED** [review]

File: packages/serverless-components/nextjs-component/src/component.ts
Comment (iiroj) [reply packages/serverless-components/nextjs-component/src/component.ts:224]: I added a custom inputs test and also updated the `next-app-experimental` to use this flag.

**Review by dphang — APPROVED** [review]

**Comment by iiroj** [comment]
@dphang I'm not familiar with the e2e test setup; would it make sense to run those before merging?

**Comment by dphang** [comment]
> @dphang I'm not familiar with the e2e test setup; would it make sense to run those before merging?

Ran it here https://github.com/serverless-nextjs/serverless-next.js/actions/runs/1744128397 (there are some issues with the experimental app which I will fix later but if it passes on the other apps it should be fine)

**Comment by iiroj** [comment]
@dphang seems like another e2e app failed, how to debug? I'll try to run this branch locally against my own site if it catches any errors.

**Comment by dphang** [comment]
Ya that app is because somehow the IAM policy got corrupted, I need to manually fix it. But it should be fine since actually your changes is primarily build-time (the apps tests are to sanity check it doesn't break existing things) and experimental now so we can add more e2e tests later.

**Comment by iiroj** [comment]
@dphang  did the tests fail because of this branch? Any way I can help debug them?

https://github.com/serverless-nextjs/serverless-next.js/actions/runs/1755195463

**Comment by dphang** [comment]
Hm I think a new minor version of nextjs yesterday is causing failures since we pick the latest one, need to look into logs and fix it but your  changes hadn't failed earlier when I ran tests. I will check why it is failing, they may have introduced some new required files.

**Pull request merged**
